qiita_db/processing_job.py

```diff
@@ -337,3 +507,466 @@ def step(self, value):
+                # the current job, so create it
+                new_job = ProcessingJob.create(
+                    user, qdb.software.Parameters.from_default_params(
+                        dflt_params, job_req_params))
+                node_to_job[n] = new_job
+
+                # Create the parent-child links in the DB
+                sql_args = [[pid, new_job.id] for pid in parent_ids]
+                qdb.sql_connection.TRN.add(sql, sql_args, many=True)
+
+            return cls._common_creation_steps(user, root_jobs, name)
+
+    @classmethod
+    def from_scratch(cls, user, parameters, name=None):
+        """Creates a new processing workflow from scratch
+
+        Parameters
+        ----------
+        user : qiita_db.user.User
+            The user creating the workflow
+        parameters : qiita_db.software.Parameters
+            The parameters of the first job in the workflow
+        name : str, optional
+            Name of the workflow. Default: generated from user's name
+
+        Returns
+        -------
+        qiita_db.processing_job.ProcessingWorkflow
+            The newly created workflow
+        """
+        job = ProcessingJob.create(user, parameters)
+        return cls._common_creation_steps(user, [job], name)
+
+    @property
+    def name(self):
+        """"The name of the workflow
+
+        Returns
+        -------
+        str
+            The name of the workflow
+        """
+        with qdb.sql_connection.TRN:
+            sql = """SELECT name
+                     FROM qiita.processing_job_workflow
+                     WHERE processing_job_workflow_id = %s"""
+            qdb.sql_connection.TRN.add(sql, [self.id])
+            return qdb.sql_connection.TRN.execute_fetchlast()
+
+    @property
+    def user(self):
+        """The user that created the workflow
+
+        Returns
+        -------
+        qdb.user.User
+            The user that created the workflow
+        """
+        with qdb.sql_connection.TRN:
+            sql = """SELECT email
+                     FROM qiita.processing_job_workflow
+                     WHERE processing_job_workflow_id = %s"""
+            qdb.sql_connection.TRN.add(sql, [self.id])
+            email = qdb.sql_connection.TRN.execute_fetchlast()
+            return qdb.user.User(email)
+
+    @property
+    def graph(self):
+        """Returns the graph of jobs that represent the workflow
+
+        Returns
+        -------
+        networkx.DiGraph
+            The graph representing the workflow
+        """
+        g = nx.DiGraph()
+        with qdb.sql_connection.TRN:
+            # Retrieve all graph workflow nodes
+            sql = """SELECT parent_id, child_id
+                     FROM qiita.get_processing_workflow_edges(%s)"""
+            qdb.sql_connection.TRN.add(sql, [self.id])
+            edges = qdb.sql_connection.TRN.execute_fetchindex()
+            if edges:
+                nodes = {jid: ProcessingJob(jid)
```

Comment:
Done - note that I still need the nodes dictionary anyway to reduce the number of DB calls when creating the ProcessingJob (otherwise it will be generated as many times as the node connectivity)